crates/refunder/src/submitter.rs

```diff
@@ -0,0 +1,227 @@
+                    } else {
+                        tracing::warn!(?err, "submission failed");
+                    }
+                }
+            }
+        } else {
+            tracing::warn!(
+                "Refund tx are not started, as the current gas price {:?} \
+                            is higher than MAX_GAS_PRICE specified {:?}",
+                max_fee_per_gas,
+                MAX_GAS_PRICE
+            );
+        }
+        Ok(())
+    }
+}
+
+fn calculate_submission_gas_price(
+    gas_price_of_last_submission: Option<U256>,
+    web3_gas_estimation: GasPrice1559,
+    newest_nonce: U256,
+    nonce_of_last_submission: Option<U256>,
+) -> Result<GasPrice> {
+    let GasPrice1559 {
+        max_fee_per_gas, ..
+    }: gas_estimation::GasPrice1559 = web3_gas_estimation;
+    // The gas price of the refund tx is the current prevailing gas price
+    // of the web3 gas estimation plus a buffer.
+    // Since we are using Eip1559 gas specification,
+    // we will only pay the buffer if it is used
+    let mut new_max_fee_per_gas = max_fee_per_gas * (1f64 + GAS_PRICE_BUFFER_IN_PERCENT / 100f64);
+    // If tx from the previous submission was not mined,
+    // we incease the gas price
+    if let Some(nonce_of_last_submission) = nonce_of_last_submission {
+        if nonce_of_last_submission == newest_nonce {
+            // Increase the gas price to last submission's
+            // gas price plus an additional 12 %
+            if let Some(previous_gas_price) = gas_price_of_last_submission {
+                new_max_fee_per_gas = f64::max(
+                    new_max_fee_per_gas,
+                    previous_gas_price.to_f64_lossy() * GAS_PRICE_BUMP,
+                );
+            }
+        }
+    }
+    let max_fee_per_gas = U256::from(new_max_fee_per_gas as u64);
+    let gas_price = GasPrice::Eip1559 {
+        max_fee_per_gas,
+        max_priority_fee_per_gas: U256::min(max_fee_per_gas, U256::from(MAX_PRIORITY_FEE_TIP)),
+    };
+    Ok(gas_price)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_calculate_submission_gas_price() {
+        // First case: previous tx was successful
+        let max_fee_per_gas = 4_000_000_000f64;
+        let web3_gas_estimation = GasPrice1559 {
+            base_fee_per_gas: 2_000_000_000f64,
+            max_fee_per_gas,
+            max_priority_fee_per_gas: 2_000_000_000f64,
+        };
+        let newest_nonce = U256::one();
+        let nonce_of_last_submission = None;
+        let gas_price_of_last_submission = None;
+        let result = calculate_submission_gas_price(
+            gas_price_of_last_submission,
+            web3_gas_estimation,
+            newest_nonce,
+            nonce_of_last_submission,
+        )
+        .unwrap();
+        let expected_result = GasPrice::Eip1559 {
+            max_fee_per_gas: U256::from(
+                (max_fee_per_gas * (1f64 + GAS_PRICE_BUFFER_IN_PERCENT / 100f64)) as u64,
```

Comment:
Maybe this constant should be specified as a factor instead of percent. `1` -> `1.01`.